### ISSUE TITLE
[Core] Validate returned client and consensus states from context calls in handlers

### ISSUE THREAD

**Issue by Farhad-Shabani** [cosmos/ibc-rs#736]
## Summary
How can IBC-rs handlers ensure that a returned `ClientState` and `ConsensusState` from a [`client_state`](https://github.com/cosmos/ibc-rs/blob/522aef1a68b80966e8e62fdacd5e3b05818f2784/crates/ibc/src/core/context.rs#L117-L118) or [`consensus_state`](https://github.com/cosmos/ibc-rs/blob/522aef1a68b80966e8e62fdacd5e3b05818f2784/crates/ibc/src/core/context.rs#L123-L130) call on the host context have valid fields? and not manipulated? Notably considering that light client developers may define any fields for their clients.

## Problem Statement
Now, we assume any returned value of true type is valid!
Though, as a general principle, whatever enters the IBC-rs boundary from the outside world, whether it's an IBC message or a datagram from the host store should be verified.

On some context calls like [`channel_end`](https://github.com/cosmos/ibc-rs/blob/522aef1a68b80966e8e62fdacd5e3b05818f2784/crates/ibc/src/core/context.rs#L204-L205), IBC-rs knows the exact returned type and its fields, allowing handlers to validate each field as necessary (e.g. [here](https://github.com/cosmos/ibc-rs/blob/9a7ab58326a89a127f9da50053256f84d4c72ef9/crates/ibc/src/core/ics04_channel/handler/chan_open_ack.rs#L103-L110)). Besides, in cases where we are unsure about the specific datagram that may be passed by IBC messages, we have provided an interface to cover the required validation step, giving hosts the option to validate fields as they see fit. (e.g. for [Signer](https://github.com/cosmos/ibc-rs/blob/522aef1a68b80966e8e62fdacd5e3b05818f2784/crates/ibc/src/core/context.rs#L261-L263))

However, in all the places (like [here](https://github.com/cosmos/ibc-rs/blob/62575ceb2a233cc78bbf7324e826eeb95ccb91a1/crates/ibc/src/core/ics02_client/handler/update_client.rs#L27-L28)) IBC-rs fetches `ClientState` and `ConsensusState` from the store, handlers do not include any checks in the validation process (rather than the client expiry or frozen status we already check!) and there is no method for clients to include their own fields' validation. 

Regarding the `ClientState`, it sounds like we should move [`validate()`](https://github.com/cosmos/ibc-rs/blob/62575ceb2a233cc78bbf7324e826eeb95ccb91a1/crates/ibc/src/clients/ics07_tendermint/client_state.rs#L147) method from ics07 and have it added under the ics02 ClientState trait. `validate()` would then be accessible to handlers. For the `ConsensusState` we can do a further investigation while working on the relevant PR.

**Comment by plafer** [comment]
> Though, as a general principle, whatever enters the IBC-rs boundary from the outside world, whether it's an IBC message or a datagram from the host store should be verified.

That's not true; only IBC datagrams (i.e. the 16 messages) need to be validated. The clients and host are considered trusted code. So we don't validate the `client_state` or `consensus_state` because trusted code is manages the store.

**Comment by Farhad-Shabani** [comment]
> The clients and host are considered trusted code. So we don't validate the client_state or consensus_state because trusted code is manages the store.

This made me think, we should explicitly communicate such assumptions somewhere in our docs. Perhaps group them under a "Security Model" section.

Though, This is not just a matter of trusting the hosts. Isn't this needed to safeguard the integrity of the IBC module against potential mess from other parts of a system? **(1)** We can’t assume hosts are fully vetted and free of bugs and **(2)** can’t be sure chains will not expose the IBC store to the other modules, meaning the store could still be compromised through non-IBC incoming transactions/messages. In the particular case of client & consensus states, their store is in access to upgrade & governance modules.
